Optional CRD Validation (#268)
operator-framework/rukpak: cmd/crdvalidator/README.md

# crdvalidator

## Summary

> A part of the core value proposition for RukPak is the safe upgrade of `CRD`s. As a result, this repository comes equipped with a "CRD Validator Webhook", `crdvalidator`, that will validate specified `CRD` upgrade events.

Reviewer: nit: could we use the full name of the resource and then use the abbreviation from then on? So CustomResourceDefinition (CRD). Also maybe a link to the upstream k8s docs on CustomResourceDefinitions could be helpful here as well?

> A part of the core value proposition for RukPak is the safe upgrade of `CRD`s. As a result, this repository comes equipped with a "CRD Validator Webhook", `crdvalidator`, that will validate specified `CRD` upgrade events.

Reviewer: I think it would also be helpful to give a background on why CRD updates are potentially risky and some of the nuances involved. Just so someone reading this immediately understands why the webhook is useful, especially if they are not already well-versed in the problem. Could be done as a follow-up though.

### How to use

> The CRD Validator webhook will ensure that any CRD upgrade is safe as long as the incoming CRD has the `core.rukpak.io/enable-safe-upgrade-validation` annotation set to `enabled` or if it is owned by RukPak via an owner reference.

Reviewer: nit: `owned by a rukpak provisioner` is more clear.

I guess we make an implicit assumption that other provisioners will implement the same ownerref pattern for objects that they manage. I think it's safe to make that assumption now but might be worth calling out in some kind of "provisioner suggested patterns" doc somewhere down the line.

Author: Yeah, we should probably create a provisioner onboarding document at some point. Probably not a huge issue now as we move closer to GA but will become important.

> The CRD Validator webhook will ensure that any CRD upgrade is safe as long as the incoming CRD has the `core.rukpak.io/enable-safe-upgrade-validation` annotation set to `enabled` or if it is owned by a RukPak provisioner.

Reviewer 2: I wonder if `safe-upgrade-validation` is misleading. I'm sure there are upgrades this webhook would deem safe that in fact are not safe (e.g. a conversion webhook in the CRD does something wrong -- it would be impossible for us to tell).

Should we have explicit annotations for each of the checks that we _are_ doing? For example: 
- `core.rukpak.io/block-removal-of-inuse-stored-version: true`
- `core.rukpak.io/block-schema-tightening: true`

Reviewer 2: Seems reasonable, since we're still on the "breaking changes are what we do" train :)

Reviewer 3: Created #289, and made a note about the uncertainly around whether it makes sense to expose this kind of configuration behavior. These kinds of configuration knobs appear to be appropriate for upstream environments, but potentially problematic in downstream environments.

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.7.0
    core.rukpak.io/safe-upgrade-validation: enabled # <--
# 
# ...
#
```

If you would like to disable validation for a CRD created by RukPak, you will need to explicitly set this annotation as `disabled` when it is added to the `Bundle`.

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.7.0
    core.rukpak.io/safe-upgrade-validation: disabled # <--
# 
# ...
#
```

### What is protected
There are three main cases `crdvalidator` is checking against with every specified `CRD` upgrade. These are where a new `CRD`:

> 1. Removes a version that the old `CRD` had and it must ensured that nothing is stored at the removed version.

Reviewer: nit: `Removes a stored version and ensures that removing the stored version does not result in data loss` for example?

                    type: array
                    minItems: 1
                    items:
                    type: string
```

With this installed onto the cluster, you go ahead and create a `CR` of a `Sample` onto the cluster. If you were to then update that `CRD`'s schema in such a way that that `CR` you created was invalidated, like so:

```yaml
#
# ...
#
spec:
    versions:
    - name: v1alpha1
        schema:
        openAPIV3Schema:
            properties:
                examplearray:
                    description: Examplearray is an array of strings to demonstrate crd validation logic.
                    type: array
                    minItems: 100
                    items:
                    type: string
```

`crdvalidator` will prevent the update from occurring if the existing `CR` does not have at least 100 items in the `examplearray` field.

## Running locally

`crdvalidator` is included with every default installation of RukPak. To run this locally, simply run the make target with a kind cluster started.

```console
$ make run
```

You can additionally run the `crdvalidator`'s e2e suite along side of RukPak's.

```console

> $ make e2e

Reviewer: Should it be `make e2e TEST='crd validation webhook'`  if they want to run those tests specifically?

Author: Hm, yeah it is probably a good idea to bring that up.

```

> If you would like to install the `crdvalidator` onto your current cluster without RukPak, you can easily do this with the appropriate make target.

Reviewer: Should we mention the cert-manager dependency? Right now we refactored the Makefile to pull the cert-manager dependency out of the install path of the individual webhook. Might be helpful to mention `make cert-manager` to install cert-manager as a prereq.

# crdvalidator

## Summary
A part of the core value proposition for RukPak is the safe upgrade of a [CustomResourceDefinition](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) (`CRD`). As a result, this repository comes equipped with a "CRD Validator Webhook" (`crdvalidator`), that will validate all `CRD` upgrades created by RukPak. This protects your `BundleInstance` pivots from having potentially dangerous effects such as data loss. 

### How to use

> `crdvalidator` will ensure that any `CRD` upgrade is safe as long as it is owned by RukPak. If you would like to disable validation for a `CRD` included in a specific `Bundle`, you will need to explicitly set the `core.rukpak.io/safe-upgrade-validation` annotation to `disabled` within that specific `Bundle`.

Reviewer: ```suggestion
`crdvalidator` will ensure that any `CRD` upgrade is safe as long as the CRD is associated with RukPak. If you would like to disable validation for a `CRD` included in a specific `Bundle`, you will need to explicitly set the `core.rukpak.io/safe-upgrade-validation` annotation to `disabled` within that specific CRD's manifest.
```

I also _really_ think we should include `crd` in the annotation key so that it's more clear what this annotation applies to. Do we feel strongly it should not be included?

Author: Sure, that's an easy fix. I'll also add Joe's "true"/"false" suggestion while I'm at it.

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    core.rukpak.io/safe-upgrade-validation: disabled # <--
```

### What is protected
There are three main cases `crdvalidator` is checking against with every specified `CRD` upgrade. These are where a new `CRD`:
1. Removes a stored version and ensures that removing the stored version does not result in data loss.
2. Changes a version that old `CRD` had and it must validate existing CRs against the new schema.
3. Adds a version that old `CRD` does not have. In this case `crdvalidator` checks if the conversion strategy is:
    - None, then ensures that existing CRs validate with new schema.
    - Webhook, then allow update (assuming webhook handles conversion correctly)

#### Update the schema that invalidates existing CRs
Say that you have installed a `CRD` onto the cluster with a Kind of `Sample` that has a single property of `examplearray`. 

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.7.0
    core.rukpak.io/safe-upgrade-validation: enabled
  creationTimestamp: null

>   name: templates.combo.io

Reviewer: nit: since in the description on L27 you reference "a CRD with a Kind of `Sample`" should the name of this CRD be something like `sample.example.io` instead?

Author: Good catch 👍🏻

#
# ...
#
spec:
    versions:
    - name: v1alpha1
        schema:
        openAPIV3Schema:
            properties:
                examplearray:
                    description: Examplearray is an array of strings to demonstrate crd validation logic.
                    type: array
                    minItems: 1
                    items:
                    type: string
```

> With this installed onto the cluster, you go ahead and create a `CR` of a `Sample` onto the cluster. If you were to then update that `CRD`'s schema in such a way that that `CR` you created was invalidated, like so:

Reviewer: should we actually show a `Sample` manifest here, or is this just for the sake of example? Is there any intent that the user would actually follow this example?

Author: That would probably make this a bit clearer and flow more naturally. I'll add that.

```yaml
#
# ...
#
spec:
    versions:
    - name: v1alpha1
        schema:
        openAPIV3Schema:
            properties:
                examplearray:
                    description: Examplearray is an array of strings to demonstrate crd validation logic.
                    type: array
                    minItems: 100
                    items:
                    type: string
```

> `crdvalidator` will prevent the update from occurring if the existing `CR` does not have at least 100 items in the `examplearray` field.

Reviewer: Might be helpful to include the specific error that the webhook returns so users can be familiar with this type of error up front.

## Running locally

`crdvalidator` is included with every default installation of RukPak. To run this locally, simply run the make target with a kind cluster started.

```console

> $ make run

Reviewer: nit: I think we decided to pull the `$` out of our bash documentation based on previous PRs

```

By default, `crdvalidator`'s test suite runs alongside RukPak's, but you can also just run `crdvalidator`'s specific tests.

```console

> $ make e2e TEST="crdvalidator"

Reviewer: nit: same here wrt removing the `$`
